src/domain/core/src/services/ingest/reader.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::backtrace::Backtrace;
+use std::path::Path;
+
+use datafusion::arrow::datatypes::Schema;
+use datafusion::prelude::*;
+use internal_error::*;
+use opendatafabric::ReadStep;
+
+///////////////////////////////////////////////////////////////////////////////
+
+/// A common interface for readers that implement support for various formats
+/// defined in the [ReadStep].
+#[async_trait::async_trait]
+pub trait Reader: Send + Sync {
+    /// Returns schema that the output will be coerced into, if such schema is
+    /// defined in the [ReadStep].
+    async fn output_schema(
+        &self,
+        ctx: &SessionContext,
+        conf: &ReadStep,
+    ) -> Result<Option<Schema>, ReadError>;
```

Comment:
Can we use this to improve UX with quick schema feedback?

Reply:
Not this function specifically, but I definitely had the schema inference for UI in mind when writing these reader types. I think DataFusion ingest is a significant step in this direction.